In [33]:
from datascience import *
import numpy as np
from math import *
from scipy import stats
import matplotlib.pyplot as plt
%matplotlib inline

## Lesson 24: Hypothesis Testing Errors & Power

Throughout this block, we have been studying hypothesis tests. We have covered the four basic steps of any hypothesis test, and we have practiced various methods for obtaining the distribution of our test statistic under the null hypothesis. 

After we have reached a conclusion (reject or fail to reject), we must consider possible errors. 

### Type I error 

Type I error is the event that we rejected the null hypothesis when the null hypothesis was actually true. Type I error is also known as a false positive. The probability of a Type I error is usually defined by the threshold used for rejection. A common threshold is 0.05. Those of you who have taken statistics before may recognize this value as $\alpha$. 

### Type II error

Type II error is the event that we failed to reject the null hypothesis when the null hypothesis was actually false. This is otherwise known as a false negative. The probability of a Type II error is harder to find and requires a more in-depth analysis of a hypothesis test. The probability of a Type II error is often given as $\beta$, and $1-\beta$ is referred to as **Power**. The power of a test is probability that we will reject the null hypothesis when we are supposed to. 

Which one of these errors is more serious? It depends on the context of the problem. 

### Example: Golf Balls

Joe has a summer job at a golf course and one of his jobs is to fish out golf balls from the water traps. He has a theory that certain types of golf ball are more likely to end up in the water than others. Let's assume there are four brands of golf ball, let's and assume that all four are used equally at this golf course. He fishes out 100 golf balls and counts each brand. He finds 30 of brand A, 30 of brand B, 20 of brand C and 20 of brand D. Conduct a hypothesis test to determine whether certain types of golf ball are more likely than others to end up in the water.

Step 1: Hypotheses

H0: P(a) = P(b) = P(c) = P(d)  
Ha: at least one P is different

Step 2: Test statistic

There are many correct answers, but let's go with sum of absolute difference between observed and expected counts under $H_0$. To do this, we need to find the expected counts. If each ball was equally likely, how many should we expected to find of each if we select 100 golf balls? 

In [34]:
# expected value = 25

Step 3: $p$-value

We need the distribution of the test statistic under $H_0$. 

In [35]:
observations  = make_array(30,30,20,20)
expected = make_array(25,25,25,25)
test_stat = sum(abs(observations-expected))
test_stat

20

In [52]:
# test statistic distribution
ts_dist = []
for i in np.arange(10000): 
    #simulation
    rs = stats.multinomial.rvs(100,(0.25,0.25,0.25,0.25),size=1)
    ts_dist = np.append(ts_dist,sum(abs(sum(rs)-25)))
# get p-value 
# ... probability that we get data equal to or more extreme than our sample assuming the null is true
np.mean(ts>=test_stat)

0.1868

Step 4: Conclude

0.1868
the p-value is high as compared to an alpha of .05 so there insufficient evidence against our null hypothesis therefore we fail to reject the null.

What kind of error could we have made in this case? 

We could have made a Type II(false negative) error.

#### Power 
Suppose that, in truth, 30% of the balls found in the water were brand A, 30% were brand B, 20% were brand C and 20% were brand D. In this case, our collected sample reflected this truth perfectly. However, our hypothesis test failed to recognize this deviation from equal proportions. We made a type II error. This is because this test has fairly low power. Use simulation to determine the power of this test. 

I am looking for the probability that I reject the null hypothesis given the true proportions laid out above. Well, first I need to figure out for what values of my test statistic I would reject $H_0$. 

In [47]:
# figure out for what values reject null
crit_value = percentile(95,ts_dist)
crit_value

24.0

Next, I need to simulate from the true population and determine how often my test statistic would have met this threshold. 

In [50]:
power = []
for i in np.arange(10000):
    rs = stats.multinomial.rvs(100,(.3,.3,.2,.2),size=1)
    power = np.append(power,sum(abs(sum(rs)-25)))
power
    # get p-value
np.mean(power>=crit_value)

0.4394

What do you think about this power? 

This is not a very reliable power. The test will only identify a false hypothesis roughly 43% of the time even though it was given to be false above. This may be due to the relatively small sample size of our data.

Repeat this power calculation, but assume Joe collects 500 balls instead of 100. Note that you will have to obtain a new critical value. What does this tell you about power and sample size?

In [60]:
# test statistic distribution
ts500_dist = []
for i in np.arange(10000): 
    #simulation
    rs = stats.multinomial.rvs(500,(0.25,0.25,0.25,0.25),size=1)
    ts500_dist = np.append(ts500_dist,sum(abs(sum(rs)-(5*25))))
# get p-value 
# ... probability that we get data equal to or more extreme than our sample assuming the null is true
crit500 = percentile(95, ts500_dist)
np.mean(ts500_dist>=crit500)

0.0617

In [ ]:
# Power = 1 - Type II error

In [65]:
power_500 = []
for i in np.arange(10000):
    rs = stats.multinomial.rvs(500,(.3,.3,.2,.2),size=1)
    power_500 = np.append(power_500,sum(abs(sum(rs)-(25*5))))
power_500
    # get power
np.mean(power_500>=crit500)

0.986

The power is a lot higher now. Power is directly related to sample size; power increases with an increase in sample size.